sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py

```diff
@@ -171,54 +175,67 @@ def _next_message_in_buffer(self):
         self._last_received_event = event_data
         return event_data
 
-    async def receive(self, batch=False, max_batch_size=300, max_wait_time=None) -> None:
+    async def _callback_coroutine(self, batch, max_batch_size, max_wait_time):
+        while self._callback_task_run:
+            try:
+                async with self._message_buffer_lock:
+                    messages = [
+                        self._message_buffer.popleft() for _ in range(min(max_batch_size, len(self._message_buffer)))
+                    ]
+                events = [EventData._from_message(message) for message in messages]
+                now_time = time.time()
+                if len(events) > 0:
+                    await self._on_event_received(events if batch else events[0])
+                    self._last_callback_called_time = now_time
+                else:
+                    if max_wait_time and (now_time - self._last_callback_called_time) > max_wait_time:
+                        # no events received, and need to callback
+                        await self._on_event_received([] if batch else None)
+                        self._last_callback_called_time = now_time
+                    # backoff a bit to avoid throttling CPU when no events are coming
+                    await asyncio.sleep(0.05)
+            except asyncio.CancelledError:
+                raise
+
+    async def _receive_coroutine(self):
         max_retries = (
             self._client._config.max_retries  # pylint:disable=protected-access
         )
-        has_not_fetched_once = True  # ensure one trip when max_wait_time is very small
-        deadline = time.time() + (max_wait_time or 0)  # max_wait_time can be None
-        while len(self._message_buffer) < max_batch_size and \
-                (time.time() < deadline or has_not_fetched_once):
-            retried_times = 0
-            has_not_fetched_once = False
-            while retried_times <= max_retries:
-                try:
-                    await self._open()
-                    await cast(ReceiveClientAsync, self._handler).do_work_async(batch=self._prefetch)
-                    break
-                except asyncio.CancelledError:  # pylint: disable=try-except-raise
-                    raise
-                except Exception as exception:  # pylint: disable=broad-except
-                    if (
+        retried_times = 0
+        while retried_times <= max_retries:
```

Comment:
This was previously wrapped within a larger `while` loop - I'm not sure we want to just exit the receive loop after max_retries?

Reply:
I have changed the receiving implementation from previous polling to current coroutine task.

the receive means differently in two implementations:

previous: the receive acts as a single call, event processor keeps calling receive,
now: receive itself a coroutine task, which do while loop (retry)

in both implementations, if retry exhausted and error raised, error would be raised. so actually the code is doing loop or raise error, it doesn't exit the loop